In [15]:
from uncertainties import ufloat
from uncertainties import unumpy as unp
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import curve_fit

g = 9.780318*((1+0.0053024*(np.sin(48)**2-0.0000058*np.sin(2*48)**2))-0.000003085*278)
l = ufloat(0.967,0.003)
a = ufloat(0.956,0.003)
d = (l/2) - (a - l)
print(d)
#print(g)
def eq1(r,q,p):
    return (2*np.pi)*(((0.25*(r**2)) + ((1/3)*(l**2)))/(p*g))**(1/2)

def mean(a):
    return sum(a)/len(a)

#print(eq1(0.005,0.956))
print(eq1(ufloat(0.0050,0.0002),ufloat(0.956,0.003),d))

T = (1.45, 1.63, 1.59, 1.63, 1.62, 1.63, 1.55, 1.66, 1.57, 1.57, 1.51,1.63, 1.49, 1.54, 1.66, 1.59, 1.63, 1.60, 1.69, 1.52, 1.67, 1.59,1.66, 1.59, 1.56, 1.56, 1.60, 1.59, 1.57, 1.58)
D = (4*np.pi**2)*(ufloat(1.109,0.005))/ufloat(502,21)
Id = (ufloat(1.82,0.02)**2*D)/(4*np.pi**2)
Iz = 0.5*ufloat(1.109,0.005)*(ufloat(0.1189,0.0001)/2)**2
print(mean(T))
#print(np.std(T))
#print(47.81/30)
#print(0.3/np.sqrt(30))
print(Iz)
print(D)
print(Id)
print(Id + Iz)
print((ufloat(2.04,0.04)**2*D)/(4*np.pi**2))
print((ufloat(4.19,0.04)*D)/(4*np.pi**2))

0.494+/-0.005
1.593+/-0.005
1.5910000000000004
0.001960+/-0.000009
0.087+/-0.004
0.00732+/-0.00035
0.00928+/-0.00035
0.0092+/-0.0005
0.0093+/-0.0004


In [16]:
ab1 = 0.015
ab2 = 0.03
ab3 = 0.045
ab4 = 0.06
T1 = ufloat(2.09,0.01)
T2 = ufloat(2.14,0.05)
T3 = ufloat(2.29,0.01)
T4 = ufloat(2.45,0.06)

I0 = ufloat(0.0093,0.0004)
m = ufloat(1.109,0.005)

I1 = I0 + m*ab1**2
I2 = I0 + m*ab2**2
I3 = I0 + m*ab3**2
I4 = I0 + m*ab4**2

M1 = (T1**2*D)/(4*np.pi**2)
M2 = (T2**2*D)/(4*np.pi**2)
M3 = (T3**2*D)/(4*np.pi**2)
M4 = (T4**2*D)/(4*np.pi**2)

print(I1,I2,I3,I4)
print(M1,M2,M3,M4)

0.0095+/-0.0004 0.0103+/-0.0004 0.0115+/-0.0004 0.0133+/-0.0004
0.0096+/-0.0004 0.0101+/-0.0006 0.0116+/-0.0005 0.0133+/-0.0009


In [18]:
Ta = 1.593
Tb = 1.59
Tc = 1.59

ua = 0.005
ub = 0.01
uc = 0.05

def t(x,y,p,q):
    return (x-y)/(np.sqrt(p**2 + q**2))

print(t(Ta,Tb,ua,ub))
print(t(Ta,Tc,ua,uc))
print(t(9.3,9.2,0.4,0.5))

0.2683281573
0.0597022314126
0.156173761889


In der Formel $I_V=\frac{V}{t}$ haben sowohl $V$ und $t$ Fehler. Wir verwenden hier also die verallgemeinerte Formel f\"ur Quotienten:
$$
\left\vert\frac{\Delta z}{z}\right\vert=\sqrt{\left(a\frac{\Delta x}{x}\right)^2+\left(b\frac{\Delta y}{y}\right)^2+\ldots}\textrm{ f\"ur }z=x^a\ y^b\ldots
$$
Hier also:
$$
\left\vert\frac{\Delta I_V}{I_V}\right\vert=\sqrt{\left(\frac{\Delta V}{V}\right)^2+\left(-1\frac{\Delta t}{t}\right)^2}
$$
Da $\frac{d^4}{l}$ aus Werten ohne vorhandenem Fehler bestehen, berechnen wir daf\"ur keinen Fehler.

Um unseren Mittelwert zu berechnen, rechnen wir ganz leicht mit 
$$
\frac{\sum_{i=1}^n I{_V_i}}{n}\label{mean}
$$
den Nominalwert, und mit
$$
s_x=\sqrt{\frac{1}{n-1}\sum_{i=1}^n(x_i-\overline{x})^2}\label{meanstd}
$$
die Standardunsichertheit dessen.

Unsere Mittelwerte der $I_V$ f\"ur jede Position werden dann gegen $\nicefrac{d^4}{l}$ aufgetragen, siehe Abbildung (\ref{Abb1}).

Da wir jedoch klar erkennen k\"onnen, dass $I{_V_4}$ mit der linearen Steigung der anderen Werte nicht \"ubereinstimmt, lassen wir diesen Wert weg und erhalten die Gerade, welche in Abbildung (\ref{Abb2}) gefunden werden kann.


Um die Steigung der Ausgleichsgeraden zu berechnen, nehmen wir folgende Formel zunutze:
$$
a=\frac{\sum x_i^2\sum y_i-\sum x_i\sum x_iy_i}{n\sum x_i^2-(\sum x_i)^2}
$$


In [11]:
def a(x,y):
    n=len(x)
    return (sum(x**2)*sum(y)-sum(x)*sum(x*y))/(n*sum(x**2)-(sum(x))**2)

print(a(np.array(darray),np.array(Iarray)))

0.041+/-0.009


Wir erhalten als Ergebnis daraus f\"ur unser $a$ einen Wert von $(0.014\pm0.011)\,\mathrm{mm}^3$

Um aus unseren Werten $\Delta p$ zu berechnen, verwenden wir
$$
\Delta p=\rho_w hg
$$

In [12]:
rho=1000
g=9.81

def Dp(h):
    return rho*h*g

print(Dp(ufloat(0.0135,0.003)))

132+/-29


Da der einzige Wert mit einem Fehler $h$ ist, rechnen wir einfach mit
$$
\Delta z=\left|{\dif f}{\dif x}\right|\Delta x\textrm{ f\"ur }z=f(x)
$$
unseren Fehler aus.
Mit $\rho_w=1000\,\frac{\mathrm{kg}}{\mathrm{m}^3}$, $g=9.81\,\frac{\mathrm{m}}{\mathrm{s}^2}$ und $h=(135\pm3)\,\mathrm{mm}$ erhalten wir als Wert $\Delta p=(132\pm29)\,$bar.

Da wir als Endergebnis $\eta$ wollen, m\"ussen wir erstmal die Gleichung (\ref{hagen}) umstellen und wir erhalten:
$$
\eta=\frac{\pi R^4\Delta p}{8I_V l}.
$$
Hier haben $\Delta p$ und $I_V$ Fehler. Wir wenden also wieder die Gleichung f\"ur Produkte an und erhalten:
$$
\left\vert\frac{\Delta\eta}{\eta}\right\vert=\sqrt{\left(\frac{\Delta\Delta p}{\Delta p}\right)^2+\left(-1\frac{\Delta I_V}{I_V}\right)^2}
$$

In [13]:
def eta(R,p,I,l):
    return np.pi*R**4*p/(8*I*l)

Dep=Dp(ufloat(0.0135,0.003))

print(eta(d1/2,Dep,I1,l1))
print(eta(d2/2,Dep,I2,l3))
print(eta(d3/2,Dep,I3,l3))
print(eta(d5/2,Dep,I5,l5))

print(mean([eta(d1/2,Dep,I1,l1),eta(d2/2,Dep,I2,l2),eta(d4/2,Dep,I4,l4),eta(d5/2,Dep,I5,l5)]))

0.076+/-0.017
0.118+/-0.027
0.088+/-0.020
0.091+/-0.020
0.081+/-0.018


Als Ergebnis f\"ur $\eta$ erhalten wir f\"ur unsere vier verwendeten Messreihen:
$\begin{itemize}
\item (0.076\pm0.017)\,\mathrm{\frac{kg}{ms}}
\item (0.118\pm0.027)\,\mathrm{\frac{kg}{ms}}
\item (0.088\pm0.020)\,\mathrm{\frac{kg}{ms}}
\item (0.091\pm0.020)\,\mathrm{\frac{kg}{ms}}
\end{itemize}
$

Nutzen wir die Formeln (\ref{mean}) und (\ref{meanstd}) um unseren Mittelwert zu bestimmen, so erhalten wir als Standardunsicherheit
\[
0.08\pm0.018\,\mathrm{\frac{kg}{ms}}
\]

Als n\"achstes betrachten wir den durchschnittlichen Fehler der Messungen und die Streuung:

In [32]:
print(mean(Iarrays))
print(np.std(Iarrayn))

NameError: name 'mean' is not defined

Wir erhalten als durchschnittlichen Fehler $0.006\,$A und als Streuung $0.570\,$A.

Anhand der Fehlerbalken ist zu erkennen, dass der Fehler mit zunehmenden $\frac{d^4}{l}$ steigt. Da die Formel f\"ur $I_V$ zu  $\frac{1}{l}$ und $\left(\frac{d}{2}\right)^2$ proportional ist, aber diese Werte keine statistischen Fehler haben, ist klar, dass dies aufgrund von systematischer Fehler der Fall sein muss. Erstaunlich ist dies nicht, da sowohl $d$ und $l$ vorgegeben worden sind ohne Fehler, wodurch sie systematische Fehler sind.

